### PR TITLE
add hazmat.FdStream

### LINKED ISSUE #829: Should we somehow expose Pipe*Stream on Unix, and if so, how?
We have `PipeReceiveStream` and `PipeSendStream` classes on Unix, that are currently not exposed publicly; they're only used internally for subprocesses.

The actual code is a pretty generic building block: they can wrap around any Unix file-descriptor that supports non-blocking mode. So they could potentially be useful for talking to [FIFOs](http://man7.org/linux/man-pages/man7/fifo.7.html), TTYs, stdio streams (#174), etc.

Should we expose them more directly? If so, how? In `trio.hazmat`, with some sort of [`os.fdopen`](https://docs.python.org/3/library/os.html#os.fdopen) equivalent?

If we do we'll want to think through the ownership semantics: should they take ownership of a passed in fd, or should they dup it?

Also, they should restore the fd's blocking/non-blocking state before closing it.

### PR DESCRIPTION
**Pull request by belm0**
Closes #829.

TODO:
 * [x] port PipeSend/ReceiveStream tests to FdStream
 * [x] update docs

### CODE REVIEW

File: trio/_unix_pipes.py [hunk 1]
```diff
@@ -157,3 +160,25 @@ def __init__(self, fd: int):
 
     def fileno(self):
         return self._fd_holder.fd
+
+
+def fd_open_send_stream(fd: int):
```
Comment (oremanj) [inline trio/_unix_pipes.py]: Name bikeshed: I feel like `open_` going at the beginning is maybe an important convention to maintain? Maybe `open_send_stream_to_fd` and `open_receive_stream_from_fd`, or even `open_stream_{to,from}_fd` if we think the send/receive-ness is obvious enough from the to/from-ness.

File: trio/_unix_pipes.py [hunk 2]
```diff
@@ -157,3 +160,25 @@ def __init__(self, fd: int):
...
+    """Returns a send stream over an os.pipe.
+
+    The pipe must support non-blocking mode.  The stream takes ownership
```
Comment (oremanj) [inline trio/_unix_pipes.py]: Elaborate what you mean here? Maybe
```
*fd* must refer to a file that is open for writing and supports non-blocking I/O (pipes and TTYs will
work, on-disk files probably not). The returned stream takes ownership of *fd*, so closing the stream
will close *fd* too. As with `os.fdopen`, you should not directly use *fd* after you have wrapped it
in a stream using this function.
```

**Comment by njsmith** [comment]
I'm trying to think of the reasons why we use `open_*` functions:

- When we can't use `__init__` because we need a different calling convention (e.g. async `open_file`, or async context manager `open_nursery`)
- When we want to hide the actual type (`open_nursery`, the old `open_cancel_scope`) – but we're moving away from this (#778)
- When it's a convenience function that does some non-trivial setup (`open_tcp_stream`)

I don't think any of these apply here though? So we should probably expose the type(s) themselves instead of having `open_*` functions.

I also think it's awkward to have separate types for send and receive here... a generic fd supports both `read` and `write`, so it's weird to have to use `StapledStream` + `dup` to read-and-write to the same fd. Plus using `dup` here is a bit tricky, because we end up with two different objects trying to fiddle with the `fcntl` flags on the same underlying file description, so it increases the risk of them stomping on each other. And we don't have `SocketSendStream` and `SocketReceiveStream` either.

So I'm thinking maybe we should merge these together into a single `trio.hazmat.FdStream`.

This would mean that on Unix `Process.stdin` and friends are all `FdStream`s, even though they're unidirectional. But I think that's OK – if someone does accidentally try to call `stdin.receive_some` or `stdout.send_all` then at runtime that will raise an exception, and we'll statically declare those attributes to be a `SendStream` and `ReceiveStream` respectively, so mypy will also flag the error.

**Comment by belm0** [comment]
> and we'll statically declare those attributes to be a SendStream and ReceiveStream respectively, so mypy will also flag the error.

I got lost here.  If `Process.stdin` is typed as a `FdStream`, how can it also be declared as a `SendStream`?

**Comment by njsmith** [comment]
`FdStream` is a subtype of `SendStream`, so we can statically type `Process.stdin` as `ReceiveStream`, and then make it a `FdStream` at runtime. (And we certainly can't statically type it as `FdStream`, because it will be something else on Windows...)

**Comment by belm0** [comment]
I've implemented `FdStream`.  Not quite sure how Process.stdio should be declared in the Unix case.

TODO:
 * [ ] port PipeSend/ReceiveStream tests to FdStream
 * [ ] Process.stdio
 * [ ] ?

**Comment by belm0** [comment]
Aside:  I believe StapledStream `send_stream` and `receive_stream` attributes should be hidden.  Anyway changing Process.stdio to `FdStream` instance in the Unix case is going to break access to them.

**Comment by njsmith** [comment]
I would expect that we'd have:

`Process.stdin`: typed as `SendStream`, on Unix actually an `FdStream`
`Process.stdout`, `Process.stderr`: typed as `ReceiveStream`, on Unix actually an `FdStream`
`Process.stdio`: `StapledStream(stdin, stdout)`

Note that on Unix we use pipes to talk to the child process, and pipes are unidirectional, so `stdio` has to be a `StapledStream`...

**Comment by njsmith** [comment]
Oh I see, you're defining `FdStream` as managing two fds, so it kind of has the `StapledStream` functionality baked into it. I was suggesting making `FdStream` manage a single fd, similar to `SocketStream`.

**Comment by belm0** [comment]
> Oh I see, you're defining `FdStream` as managing two fds

It did make it easy to ensure only one holder when receive and send streams were the same.  Any ideas for that given the single-fd FdStream?

**Comment by njsmith** [comment]
If the send and receive streams are the same, then just... Use one `FdStream`? I don't think I'm understanding the question.

**Comment by belm0** [comment]
OK-- I think the API is settled, so it's a matter of tests and docs.

I didn't do anything to preemptively check for accidental send_all / receive_some on the wrong mode at runtime.  I'm assuming it's fine to defer to OSError.

**Comment by belm0** [comment]
another thing I wasn't quite sure about:  in the implementation I've used separate task conflict detectors for send & receive

NOTE: please squash on merge

File: trio/_unix_pipes.py [hunk 3]
```diff
@@ -33,9 +33,11 @@ def __init__(self, fd: int):
...
             raise TypeError("file descriptor must be an int")
         self.fd = fd
-        # Flip the fd to non-blocking mode
-        flags = fcntl.fcntl(self.fd, fcntl.F_GETFL)
-        fcntl.fcntl(self.fd, fcntl.F_SETFL, flags | os.O_NONBLOCK)
+        # Store original state, and ensure non-blocking mode is enabled
+        self._original_flags = fcntl.fcntl(self.fd, fcntl.F_GETFL)
+        fcntl.fcntl(
+            self.fd, fcntl.F_SETFL, self._original_flags | os.O_NONBLOCK
+        )
```
Comment (njsmith) [inline trio/_unix_pipes.py]: Let's use `os.get_blocking` and `os.set_blocking`. Advantages:

- Less fcntl line noise
- In the unlikely event that someone else is fiddling with other unrelated fcntl flags while the `FdStream` is open, we won't inadverdantly overwrite their changes when we reset the blocking state

File: trio/_unix_pipes.py [hunk 4]
```diff
@@ -53,6 +55,7 @@ def _raw_close(self):
             return
         fd = self.fd
         self.fd = -1
+        fcntl.fcntl(fd, fcntl.F_SETFL, self._original_flags)
```
Comment (njsmith) [inline trio/_unix_pipes.py]: Likewise here

File: trio/_unix_pipes.py [hunk 5]
```diff
@@ -65,21 +68,38 @@ def __del__(self):
...
     def __init__(self, fd: int):
         self._fd_holder = _FdHolder(fd)
-        self._conflict_detector = ConflictDetector(
-            "another task is using this pipe"
+        self._send_conflict_detector = ConflictDetector(
+            "another task is using this stream for send"
+        )
+        self._receive_conflict_detector = ConflictDetector(
+            "another task is using this stream for receive"
         )
```
Comment (njsmith) [inline trio/_unix_pipes.py]: Yep, this is the right way to do it.

File: docs/source/reference-hazmat.rst [hunk 6]
```diff
@@ -174,6 +174,40 @@ All environments provide the following functions:
...
+or processes are using file descriptors that are related through `os.dup`
+or inheritance across `os.fork` to the one that Trio is using, they are
+unlikely to be prepared to have non-blocking I/O semantics suddenly
+thrust upon them.  For example, you can use ``FdStream(os.dup(0))`` to
+obtain a stream for reading from standard input, but it is only safe to
+do so with heavy caveats: your stdin must not be shared by any other
+processes and you must not make any calls to synchronous methods of
+`sys.stdin` until the stream returned by `FdStream` is closed. See
+`issue #174 <https://github.com/python-trio/trio/issues/174>`__ for a
+discussion of the challenges involved in relaxing this restriction.
```
Comment (njsmith) [inline docs/source/reference-hazmat.rst]: Maybe this should be merged into the class docstring?

**Comment by belm0** [comment]
> the only reason `_FdHolder` exists is to factor out some common code from `PipeSendStream` and `PipeReceiveStream`. Now that we're merging those two classes into one, it's no longer really serving any purpose, and we could merge it back into `FdStream`.

I'm inclined to keep the separation as it keeps FdStream readable.  FdHolder could have its own test.

**Comment by njsmith** [comment]
I made a few trivial wording tweaks, and otherwise looks great! Feel free to merge once CI passes, in case I forget...

**Comment by njsmith** [comment]
Actually, I take that back :-). The coverage report is saying that `from ._unix_pipes import FdStream` always succeeds, even on Windows: https://codecov.io/gh/python-trio/trio/pull/1129/src/trio/hazmat.py?before=trio/hazmat.py

Which makes sense, because there's nothing in `trio._unix_pipes` that depends on platform-specific stuff at import time. There used to be, when we were importing `fcntl`, but not anymore. So I guess right now we're exporting `trio.hazmat.FdStream` on Windows. But of course if anyone tries to use it on Windows then it will just crash.

(Side note: it's too bad the tests don't catch this... I guess they're doing `skipif(not os.name == "posix")`, not `skipif(not hasattr(trio.hazmat, "FdStream")`?)

The obvious solution would be to make it `if os.name == "posix": from ._unix_pipes import FdStream`, but from your commit comments, I guess jedi didn't like that. Which is irritating :-/. Annoyingly, mypy doesn't understand it either: https://mypy.readthedocs.io/en/latest/common_issues.html#version-and-platform-checks

One option would be to use `if os.name == "posix"`, and accept that jedi won't autocomplete correctly, and hack the jedi test to add a special-case exception

Another option, which is kind of a silly hack, but I guess would work: at the top of `_unix_pipes.py`, do `from os import get_blocking, set_blocking`, which will raise `ImportError` on Windows. Of course this would need a comment :-)

**Comment by belm0** [comment]
I'm now manually raising ImportError from `_unix_pipes.py`, and added coverage for this.

**Comment by njsmith** [comment]
Looks great, waiting for CI now...

**Pull request merged**